src/runtime_intrinsics.c

```diff
@@ -60,21 +60,411 @@ JL_DLLEXPORT jl_value_t *jl_pointerset(jl_value_t *p, jl_value_t *x, jl_value_t
     jl_value_t *ety = jl_tparam0(jl_typeof(p));
     if (ety == (jl_value_t*)jl_any_type) {
         jl_value_t **pp = (jl_value_t**)(jl_unbox_long(p) + (jl_unbox_long(i)-1)*sizeof(void*));
-        *pp = x;
+        *pp = x; // TODO: change alignment assumption?
     }
     else {
         if (!jl_is_datatype(ety))
             jl_error("pointerset: invalid pointer");
+        if (jl_typeof(x) != ety)
+            jl_type_error("pointerset", ety, x);
         size_t elsz = jl_datatype_size(ety);
         size_t nb = LLT_ALIGN(elsz, jl_datatype_align(ety));
         char *pp = (char*)jl_unbox_long(p) + (jl_unbox_long(i)-1)*nb;
+        memcpy(pp, x, elsz);
+    }
+    return p;
+}
+
+#ifdef _P64 // && defined(UINT128MAX)
+#define MAX_POINTERATOMIC_SIZE 16
+typedef __uint128_t uint128_t;
+#else
+#define MAX_POINTERATOMIC_SIZE 8
+//typedef struct _uint128_t {
+//    uint64_t a[2];
+//} uint128_t;
+#endif
+
+#if MAX_ATOMIC_SIZE > MAX_POINTERATOMIC_SIZE
+#error MAX_ATOMIC_SIZE too large
+#endif
+
+
+JL_DLLEXPORT jl_value_t *jl_atomic_new_bits(jl_value_t *dt, const char *data)
```

Comment:
Maybe move this next to `jl_new_bits`? I think this file should only have the entry points for the intrinsics themselves.